File: Utils/StringTransform.swift
import Foundation

// MARK: - Namespace: StringTransform
enum StringTransform {
    static func sanitize(_ s: String?) -> String {
        guard let str = s else { return "" }
        let trimmed = str.trimmingCharacters(in: .whitespaces)
        guard !trimmed.isEmpty else { return "" }

        let stem = trimmed.split(separator: ".").first?.uppercased() ?? ""
        if Config.reservedWindowsNames.contains(String(stem)) {
            return "_\(trimmed)_"
        }

        return trimmed.replacing(Config.Regex.invalidChars, with: "")
    }
    static func safeUTF8Truncate(_ string: String, maxByteLen: Int) -> String {
        if string.utf8.count <= maxByteLen { return string }
        var result = ""; var currentBytes = 0
        for char in string {
            let charBytes = char.utf8.count
            if currentBytes + charBytes > maxByteLen { break }
            result.append(char); currentBytes += charBytes
        }
        return result
    }
    static func truncatePathComponent(_ component: String, maxLen: Int = Config.Limits.maxPathComponentLength) -> (String, Bool) {
        if component.utf8.count <= maxLen { return (component, false) }
        var stem = component; var ext = ""
        if let dotIndex = component.lastIndex(of: "."), dotIndex != component.startIndex {
            stem = String(component[..<dotIndex]); ext = String(component[dotIndex...])
        }
        let extLen = ext.utf8.count; let suffix = "..."; let suffixLen = suffix.utf8.count
        let maxStemLen = maxLen - extLen - suffixLen
        if maxStemLen <= 0 {
            let safeStr = safeUTF8Truncate(component, maxByteLen: maxLen)
            Console.warning("Path truncated (ext too long): \(component) -> \(safeStr)")
            return (safeStr, true)
        }
        let safeStem = safeUTF8Truncate(stem, maxByteLen: maxStemLen)
        let finalStr = safeStem + suffix + ext
        Console.warning("Path truncated: \(component) -> \(finalStr)")
        return (finalStr, true)
    }
    private static func isValidDate(year: Int, month: Int, day: Int) -> Bool {
        var components = DateComponents(); components.year = year; components.month = month; components.day = day
        return Calendar.current.date(from: components) != nil
    }
    private static func validateAndFormatDate(year: Int, month: Int, day: Int) -> String? {
        guard (Config.Limits.yearMin...Config.Limits.yearMax).contains(year), (1...12).contains(month), (1...31).contains(day), isValidDate(year: year, month: month, day: day) else { return nil }
        return String(format: "%04d%02d%02d", year, month, day)
    }
    static func formatDateString(_ value: String) -> String {
        var strValue = value.trimmingCharacters(in: .whitespaces)
        guard !strValue.isEmpty else { return "" }

        // Remove date-time suffix
        strValue = strValue.replacing(Config.Regex.dateTimeSuffix, with: "")

        // If all digits and not 8 digits, early return sanitize
        let isAllDigits = strValue.unicodeScalars.allSatisfy { CharacterSet.decimalDigits.contains($0) }
        if isAllDigits && strValue.count != 8 {
            return sanitize(strValue)
        }

        // Try date separator split
        let parts = strValue.components(separatedBy: CharacterSet(charactersIn: "-/.")).filter { !$0.isEmpty }
        if parts.count == 3 {
            // YYYY-MM-DD
            if parts[0].count == 4, let y = Int(parts[0]), let m = Int(parts[1]), let d = Int(parts[2]),
               let formatted = validateAndFormatDate(year: y, month: m, day: d) {
                return formatted
            }
            // MM-DD-YYYY or DD-MM-YYYY
            if parts[2].count == 4, let y = Int(parts[2]) {
                if let m = Int(parts[0]), let d = Int(parts[1]),
                   let formatted = validateAndFormatDate(year: y, month: m, day: d) {
                    return formatted
                }
                if let d = Int(parts[0]), let m = Int(parts[1]),
                   let formatted = validateAndFormatDate(year: y, month: m, day: d) {
                    return formatted
                }
            }
        }

        // Chinese date pattern search
        let nsStr = strValue as NSString
        let chinesePattern = "(\\d{4})年(\\d{1,2})月(\\d{1,2})日"
        if let cnRegex = try? NSRegularExpression(pattern: chinesePattern),
           let cnMatch = cnRegex.firstMatch(in: strValue, range: NSRange(location: 0, length: nsStr.length)) {
            let yRange = cnMatch.range(at: 1)
            let mRange = cnMatch.range(at: 2)
            let dRange = cnMatch.range(at: 3)
            if let y = Int(nsStr.substring(with: yRange)),
               let m = Int(nsStr.substring(with: mRange)),
               let d = Int(nsStr.substring(with: dRange)),
               let formatted = validateAndFormatDate(year: y, month: m, day: d) {
                return formatted
            }
        }

        // 8-digit date
        if (try? Config.Regex.eightDigit.wholeMatch(in: strValue)) != nil {
            if let y = Int(strValue.prefix(4)),
               let m = Int(strValue[strValue.index(strValue.startIndex, offsetBy: 4)..<strValue.index(strValue.startIndex, offsetBy: 6)]),
               let d = Int(strValue.suffix(2)),
               let formatted = validateAndFormatDate(year: y, month: m, day: d) {
                return formatted
            }
        }

        // Fallback: sanitize
        return sanitize(strValue)
    }
}

File: python/exconverter/string_transform.py
"""String cleaning, UTF-8-safe truncation, and date formatting
(port of Utils/StringTransform.swift).
"""

from __future__ import annotations

import datetime
import re

from . import config, console

_CHINESE_DATE_PATTERN = re.compile(r"(\d{4})年(\d{1,2})月(\d{1,2})日")


def sanitize(s: str | None) -> str:
    """Trim, drop illegal filename chars, and guard Windows reserved names.

    Mirrors Swift: trim whitespace; if empty return ""; if the component before
    the first "." uppercased is a Windows reserved name, wrap the *trimmed*
    original in underscores; otherwise strip illegal characters.
    """
    if s is None:
        return ""
    trimmed = s.strip()
    if not trimmed:
        return ""

    stem = next((part for part in trimmed.split(".") if part), "").upper()
    if stem in config.reserved_windows_names:
        return f"_{trimmed}_"

    return config.Regex.invalid_chars.sub("", trimmed)


def safe_utf8_truncate(string: str, max_byte_len: int) -> str:
    """Truncate to a UTF-8 byte budget without splitting a code point."""
    if len(string.encode("utf-8")) <= max_byte_len:
        return string
    result = ""
    current_bytes = 0
    for ch in string:
        ch_bytes = len(ch.encode("utf-8"))
        if current_bytes + ch_bytes > max_byte_len:
            break
        result += ch
        current_bytes += ch_bytes
    return result


def truncate_path_component(
    component: str, max_len: int = config.Limits.max_path_component_length
) -> tuple[str, bool]:
    """Truncate a path component to a byte budget, keeping its extension."""
    if len(component.encode("utf-8")) <= max_len:
        return component, False

    stem = component
    ext = ""
    # Swift uses lastIndex(of: ".") != startIndex
    dot_index = component.rfind(".")
    if dot_index > 0:
        stem = component[:dot_index]
        ext = component[dot_index:]

    ext_len = len(ext.encode("utf-8"))
    suffix = "..."
    suffix_len = len(suffix.encode("utf-8"))
    max_stem_len = max_len - ext_len - suffix_len

    if max_stem_len <= 0:
        safe = safe_utf8_truncate(component, max_byte_len=max_len)
        console.warning(f"Path truncated (ext too long): {component} -> {safe}")
        return safe, True

    safe_stem = safe_utf8_truncate(stem, max_byte_len=max_stem_len)
    final = safe_stem + suffix + ext
    console.warning(f"Path truncated: {component} -> {final}")
    return final, True


def _is_valid_date(year: int, month: int, day: int) -> bool:
    try:
        datetime.date(year, month, day)
        return True
    except ValueError:
        return False


def _validate_and_format_date(year: int, month: int, day: int) -> str | None:
    if not (config.Limits.year_min <= year <= config.Limits.year_max):
        return None
    if not (1 <= month <= 12):
        return None
    if not (1 <= day <= 31):
        return None
    if not _is_valid_date(year, month, day):
        return None
    return f"{year:04d}{month:02d}{day:02d}"


def format_date_string(value: str) -> str:
    """Normalize many date shapes to YYYYMMDD (port of formatDateString)."""
    str_value = value.strip()
    if not str_value:
        return ""

    # Remove a trailing date-time suffix, e.g. " 12:30:45" or " 123456".
    str_value = config.Regex.date_time_suffix.sub("", str_value)

    # All digits but not 8 -> not a date; sanitize and return early.
    is_all_digits = str_value.isdecimal()
    if is_all_digits and len(str_value) != 8:
        return sanitize(str_value)

    # Try splitting on date separators - / .
    parts = [p for p in re.split(r"[-/.]", str_value) if p]
    if len(parts) == 3:
        # YYYY-MM-DD
        if len(parts[0]) == 4:
            try:
                y, m, d = int(parts[0]), int(parts[1]), int(parts[2])
            except ValueError:
                y = m = d = None
            if y is not None:
                formatted = _validate_and_format_date(y, m, d)
                if formatted is not None:
                    return formatted
        # MM-DD-YYYY or DD-MM-YYYY
        if len(parts[2]) == 4:
            try:
                y = int(parts[2])
                m, d = int(parts[0]), int(parts[1])
            except ValueError:
                y = None
                m = d = None
            if y is not None:
                formatted = _validate_and_format_date(y, m, d)
                if formatted is not None:
                    return formatted
                # Try DD-MM-YYYY
                formatted = _validate_and_format_date(y, d, m)
                if formatted is not None:
                    return formatted

    # Chinese date pattern — searches anywhere in the string (Swift uses
    # NSRegularExpression.firstMatch, not an anchored match).
    m = _CHINESE_DATE_PATTERN.search(str_value)
    if m:
        try:
            y, mo, day = int(m.group(1)), int(m.group(2)), int(m.group(3))
        except ValueError:
            y = mo = day = None
        if y is not None:
            formatted = _validate_and_format_date(y, mo, day)
            if formatted is not None:
                return formatted

    # 8-digit date YYYYMMDD
    if config.Regex.eight_digit.fullmatch(str_value):
        try:
            y = int(str_value[:4])
            mo = int(str_value[4:6])
            day = int(str_value[6:8])
        except ValueError:
            y = mo = day = None
        if y is not None:
            formatted = _validate_and_format_date(y, mo, day)
            if formatted is not None:
                return formatted

    return sanitize(str_value)
